### PR TITLE
update for adding hashing and sorting operators to classes

### PR DESCRIPTION
**Pull request by vermavineet-google**
Allows the statements to be sorted when kept in lists and other datatypes and also useful in pandas dataframes

### CODE REVIEW

**Comment by evhub** [comment]
Thanks for the PR! Happy to add the hashing and comparison behavior, but I don't want to rename the project to `pxprover`—if you want to clean up the PR and remove the name change, I'll merge this, otherwise I'll just copy your code into a separate commit.

**Comment by vermavineet-google** [comment]
yikes...i made some changes in my py repo...will revert changes and share

**Review by vineetvermait — COMMENTED** [review]
reverted name changes

**Comment by evhub** [comment]
Thanks—this looks a lot better now. Still looks like there's a bunch of code duplication, though—inheriting `__hash__` from base classes behaves a bit weirdly, so duplication there is fine, but all the comparison operations should probably be factored out into base classes.

**Comment by vineetvermait** [comment]
Updated as per comments...also added capability to copy, was getting errors earlier

File: pyprover-source/constants.coco [hunk 1]
```diff
@@ -19,7 +19,7 @@ def first_encodeable(symbols):
 
 # Installation:
 
-version = "0.5.4"
+version = "0.5.5"
```
Comment (evhub) [inline pyprover-source/constants.coco]: Don't change this—I'll bump it later after I've tested everything.

File: pyprover-source/logic.coco [hunk 2]
```diff
@@ -113,50 +113,22 @@ class Expr:
 
 class Top(Expr):
     """True"""
-    __slots__ = ()
+    __slots__ = ('elems')
```
Comment (evhub) [inline pyprover-source/logic.coco]: This doesn't work—you need a trailing comma to make it a tuple.

File: pyprover-source/logic.coco [hunk 3]
```diff
@@ -113,50 +113,22 @@ class Expr:
...
-
-    def __le__(self, other) -> int:
-        return str(self) <= str(other)
 
 top = true = Top()
 
 class Bot(Expr):
     """False"""
-    __slots__ = ()
+    __slots__ = ('elems')
```
Comment (evhub) [inline pyprover-source/logic.coco]: Ditto.

File: pyprover-source/logic.coco [hunk 4]
```diff
@@ -113,50 +113,22 @@ class Expr:
 
 class Top(Expr):
     """True"""
-    __slots__ = ()
+    __slots__ = ('elems')
     def __eq__(self, other) = isinstance(other, Top)
     def __repr__(self) = "top"
     def __str__(self) = top_sym
     def __bool__(self) = True
-    def __hash__(self) -> int:
```
Comment (evhub) [inline pyprover-source/logic.coco]: It looks like all of this stuff was removed, but not actually added back anywhere.

**Review by evhub — CHANGES_REQUESTED** [review]

**Comment by vineetvermait** [comment]
Updated and removed the copy thing.

File: pyprover-source/logic.coco (hunk 4 above)
Comment (vineetvermait) [reply pyprover-source/logic.coco]: only the stuff that i added for hash and all is removed

**Comment by evhub** [comment]
Looks good; thanks!

**Pull request merged**